proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,820 @@
+- All clients are only allowed
+  to [use `code` as the `response_type`](https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationExamples)
+  at the authorization endpoint.
+- All clients are only allowed to use the default or specify `query` as
+  the [`response_mode`](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) at the authorization
+  endpoint. This excludes clients from using `form_post`. We could consider allowing `form_post` in the future if it is
+  desired.
+- Clients are not allowed to use JWT-based client auth. This could potentially be added as a feature in the future.
+
+##### Configuring clients via an aggregated API server (alternative data model to CRD)
+
+CRDs are convenient to use, however, they have one core limitation - they cannot represent non-CRUD semantics.  For
+example, the existing token credential request API could not be implemented using a CRD - it processes an incoming token
+and returns a client certificate without writing any data to etcd.  Aggregated APIs have no such limitation.
+
+An example of how the aggregated API would define a client:
+
+```yaml
+apiVersion: oauth.supervisor.pinniped.dev/v1alpha1
+kind: OIDCClient
+metadata:
+  name: my-webapp-client
+  namespace: pinniped-supervisor
+spec:
+  allowedRedirectURIs:
+    - https://my-webapp.example.com/callback
+  allowedGrantTypes:
+    - authorization_code
+    - refresh_token
+    - urn:ietf:params:oauth:grant-type:token-exchange
+  allowedScopes:
+    - openid
+    - offline_access
+    - pinniped:request-audience
+    - groups
+```
+
+Some key differences from the CRD based approach:
+
+- The `.status` field is omitted - validation is performed statically on every `create` and `update` operation - the
+  client object is always guaranteed to be valid.  The admin is notified directly via a validation error if they attempt
+  to create an incorrect object (for example, with `metadata.name` set to a value that includes a `:`).  Similarly, if
+  the `.spec.allowed*` fields include invalid data, the object will be rejected.  Consumers of this API, such as our
+  controllers, can rely on the data being valid, which simplifies their logic.
+- `.spec.secretNames` has been removed as the client secret is generated on the server after the client has already been
+  created.  This is handled via a `create` operation to the `secret` subresource (this is modeled after the Kubernetes
+  service account API's `token` subresource).  Just as the SA `token` subresource uses the `TokenRequest` API as the
+  input and output schema, the `secret` subresource would use a `SecretRequest` API as the input and output schema.
+  The response from this API is the only time that the plaintext client secret is made available.  Since rotation is not
+  part of this proposal, the `secret` API can only be invoked once per client - all subsequent requests will fail.  A
```

Comment:
I don't think "rotation is not a part of this proposal" is an accurate representation of the non goal

> Management (ie creation & rotation) of client credentials on the operator's behalf. 

The intention of that goal exclude automation credential rotations as part of this first effort. Enabling the operator to be able to manually implement a credential rotation without disrupting a client's ability to function is, I think, a desirable outcome.